packages/core/database/lib/entity-manager/relations-orderer.js

```diff
@@ -1,6 +1,106 @@
 'use strict';
 
 const _ = require('lodash/fp');
+const { InvalidRelationError } = require('../errors');
+/**
+ * When connecting relations, the order you connect them matters.
+ *
+ * Example, if you connect the following relations:
+ *   { id: 5, position: { before: 1 } }
+ *   { id: 1, position: { before: 2 } }
+ *   { id: 2, position: { end: true } }
+ *
+ * Going through the connect array, id 5 has to be connected before id 1,
+ * so the order of id5 = id1 - 1. But the order value of id 1 is unknown.
+ * The only way to know the order of id 1 is to connect it first.
+ *
+ * This function makes sure the relations are connected in the right order:
+ *   { id: 2, position: { end: true } }
+ *   { id: 1, position: { before: 2 } }
+ *   { id: 5, position: { before: 1 } }
+ *
+ */
+const sortConnectArray = (connectArr, initialArr = [], strictSort = true) => {
+  const sortedConnect = [];
+  // Boolean to know if we have to recalculate the order of the relations
+  let needsSorting = false;
+  // Map to validate if relation is already in sortedConnect or DB.
+  const relInArray = initialArr.reduce((acc, rel) => ({ ...acc, [rel.id]: true }), {});
+  // Map to store the first index where a relation id is connected
+  const firstSeen = {};
+  // Map to validate if connect relation has already been computed
+  const computedIdx = {};
+
+  connectArr.forEach((rel, idx) => {
+    // If adjacent relation is not in the database or seen yet in the connect array
+    // then we need to sort the connect array
+    const adjacentRelId = rel.position?.before || rel.position?.after;
+    if (!relInArray[adjacentRelId] && !firstSeen[adjacentRelId]) needsSorting = true;
+    // Populate firstSeen
+    if (!(rel.id in firstSeen)) firstSeen[rel.id] = idx;
+  });
+
+  // If we don't need to sort the connect array, we can return it as is
+  if (!needsSorting) return connectArr;
+
+  // Add relation to sortedConnect and mark it as seen
+  const pushRelation = (rel) => {
+    sortedConnect.push(rel);
+    relInArray[rel.id] = true;
+  };
+
+  // Recursively compute in which order the relation should be connected
+  const computeRelation = (rel, idx, relationsSeenInBranch) => {
```

Comment:
Can this be in a separate function declaration to reduce the complexity of the sortConnectArray function? 🤔 
It looks like it is using a lot of closure.

Reply:
I think it is not possible the way it is defined.
It needs to mutate the relInArray and sortedConnect.

We could split it a way to make that possible, but I do not find a way that's more readable than this one. Do you have any in mind that could improve it?

Reply:
> `computeRelation` needs to mutate the `relInArray` and `sortedConnect`.

Maybe we should try to not mutate anything. That we could easily split functions ? 🤔